File: data-pipeline-flink/dp-telemetry-reader/src/main/java/org/sunbird/dp/reader/ParentListOfMap.java
package org.sunbird.dp.reader;

import org.sunbird.dp.util.Logger;
import sun.reflect.generics.reflectiveObjects.NotImplementedException;

import java.util.List;
import java.util.Map;

// import org.ekstep.ep.samza.core.Logger;

class ParentListOfMap implements ParentType {
    //TODO#: Make this class more genic.
    static Logger LOGGER = new Logger(ParentListOfMap.class);
    List<Map<String, Object>> list;
    String childKey;

    ParentListOfMap(List<Map<String, Object>> list, String childKey) {
        this.list = list;
        this.childKey = childKey;
    }

    @Override
    public <T> T readChild() {
        if (list == null) {
            return null;
        }
        for (Object itemsObject : list) {
            if (!(itemsObject instanceof Map)) {
                continue;
            }
            Map<String, Object> items = (Map<String, Object>) itemsObject;
            if (items.containsKey(childKey)) {
                Object o = items.get(childKey);
                if (o instanceof List && ((List) o).size() > 0) {
                    return (T) ((List) o).get(0);
                }
            }
        }
        return null;
    }

    @Override
    public void addChild(Object value) {
        throw new NotImplementedException();
    }
}

File: data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/reader/ParentListOfMap.scala
package org.sunbird.dp.reader

//import org.sunbird.dp.util.Logger

import sun.reflect.generics.reflectiveObjects.NotImplementedException

import java.util.List

import java.util.Map

import ParentListOfMap._

//remove if not needed
import scala.collection.JavaConversions._

object ParentListOfMap {

  //TODO#: Make this class more genic.
  //var LOGGER: Logger = new Logger(classOf[ParentListOfMap])

}

class ParentListOfMap(var list: List[Map[String, Any]], var childKey: String)
  extends ParentType {

  override def readChild[T]: Option[T] = {
    if (list == null) {
      null
    }
    for (itemsObject <- list) {
      if (!(itemsObject.isInstanceOf[Map[_, _]])) {
        //continue
      }
      val items: Map[String, Any] = itemsObject.asInstanceOf[Map[String, Any]]
      if (items.containsKey(childKey)) {
        val o: AnyRef = items.get(childKey).asInstanceOf[AnyRef]
        if (o.isInstanceOf[List[_]] && o.asInstanceOf[List[_]].size > 0) {
          Some(o.asInstanceOf[List[_]].get(0).asInstanceOf[T])
        }
      }
    }
    null
  }

  override def addChild(value: Any): Unit = {
    throw new NotImplementedException()
  }

}

// import org.ekstep.ep.samza.core.Logger;
